### PR TITLE
Add ADC driver function for imxrt

### PR DESCRIPTION
**Pull request by HarrisonChoi750405**

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [7ce446f812f18dcdd3cc3699c520bc7190a57e63] -

**Comment by seinfra** [comment]
Target : [7ce446f812f18dcdd3cc3699c520bc7190a57e63] -

**Comment by seinfra** [comment]
Target : [d52898029950c9c3ca4e6f7aed1a386b954ba751] -

**Comment by seinfra** [comment]
Target : [d52898029950c9c3ca4e6f7aed1a386b954ba751] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [89ae9929c4b1420547f1ef77e79aade64da01683] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [89ae9929c4b1420547f1ef77e79aade64da01683] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fba3c5215ef3809313e3ff8e54ba77618657b5b9] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [fba3c5215ef3809313e3ff8e54ba77618657b5b9] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d1efad8c2e5bf56452ea07893e7370c235997d57] -

**Comment by seinfra** [comment]
Target : [d1efad8c2e5bf56452ea07893e7370c235997d57] -

**Comment by seinfra** [comment]
Target : [5b3e80068c6de491942655aef067d27b3ecafec5] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [5b3e80068c6de491942655aef067d27b3ecafec5] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [14f42a5855f2ded338b082c20c08ff9a43610450] -

**Comment by seinfra** [comment]
Target : [14f42a5855f2ded338b082c20c08ff9a43610450] -

File: apps/examples/testcase/ta_tc/systemio/utc/utc_adc.c [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	utc_systemio_iotbus_adc_start_n();
+
+	utc_systemio_iotbus_adc_stop_p();
+	utc_systemio_iotbus_adc_stop_n();
+
+	utc_systemio_iotbus_adc_deinit_n();
+	utc_systemio_iotbus_adc_deinit_p();
+
+	return 0;
+}
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/systemio/utc/utc_adc.c]: Let's remove this

File: os/arch/arm/src/imxrt/imxrt_adc.c [hunk 2]
```diff
@@ -0,0 +1,813 @@
...
+#include <tinyara/irq.h>
+#include <tinyara/wqueue.h>
+#include <tinyara/wdog.h>
+#include <tinyara/analog/adc.h>
+#include <tinyara/analog/ioctl.h>
+
+#include "imxrt_clock.h"
+#include "imxrt_adc.h"
+
+#ifdef CONFIG_ANALOG
```
Comment (sunghan-chang) [inline os/arch/arm/src/imxrt/imxrt_adc.c]: CONFIG_ANALOG for driver, not for arch. Let's revert this.

File: os/arch/arm/src/imxrt/imxrt_adc.c [hunk 3]
```diff
@@ -0,0 +1,813 @@
...
+
+//Depends on TICK; Default is 100ms(10ms*10)!
+#define IMXRT_ADC_DEFAULT_PEROID  10
+#define IMXRT_ADC_MIN_PEROID    1
+#define IMXRT_ADC_MAX_PEROID    1000000
+
+#define DEBUG_ADC	0
+#undef adcinfo
+#if defined(DEBUG_ADC) && DEBUG_ADC > 0
+#define adcinfo(format, ...)   printf(format, ##__VA_ARGS__)
```
Comment (sunghan-chang) [inline os/arch/arm/src/imxrt/imxrt_adc.c]: This is not allowed. Let's use existed `avdbg`.

**Comment by seinfra** [comment]
Target : [db6fd0ade96f5cb9b8e1038673e53cb66153af0c] - Code Rule Check Result:
os/arch/arm/src/imxrt/imxrt_adc.c:92: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
os/arch/arm/src/imxrt/imxrt_adc.c:92: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxV)
os/arch/arm/src/imxrt/imxrt_adc.c:168: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:275: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:277: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:281: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:282: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:282: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:283: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:283: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:284: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:288: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:289: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:292: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:293: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:294: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:294: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:295: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:295: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:296: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:296: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:297: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:297: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:298: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:301: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:302: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:302: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:303: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:304: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:305: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:305: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:306: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:309: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:313: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:314: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/imxrt/imxrt_adc.c:314: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:315: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:318: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:366: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/imxrt/imxrt_adc.c:520: ERROR: [BRC_M_SMT] that open brace { should be on the previous line
os/arch/arm/src/imxrt/imxrt_adc.c:524: ERROR: [BRC_M_SMT] that open brace { should be on the previous line
os/arch/arm/src/imxrt/imxrt_adc.c:524: ERROR: [BRC_M_SMT] else should follow close brace '}'
os/arch/arm/src/imxrt/imxrt_adc.c:577: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:578: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:647: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/imxrt/imxrt_adc.c:666: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [db6fd0ade96f5cb9b8e1038673e53cb66153af0c] - Code Rule Check (C++) OK.

File: os/arch/arm/src/imxrt/imxrt_adc.c (hunk 2 above)
Comment (Hyunjun85) [reply os/arch/arm/src/imxrt/imxrt_adc.c]: Reverted this commit.

**Comment by seinfra** [comment]
Target : [019e8faa3eb7362a5723b964011a67a67c5f5c09] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [019e8faa3eb7362a5723b964011a67a67c5f5c09] - Code Rule Check Result:
os/arch/arm/src/imxrt/imxrt_adc.c:520: ERROR: [BRC_M_SMT] that open brace { should be on the previous line

**Comment by seinfra** [comment]
Target : [19f5d0f43c647ed28e24f889a18bedfc64872e53] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [19f5d0f43c647ed28e24f889a18bedfc64872e53] - Code Rule Check Result:
os/arch/arm/src/imxrt/imxrt_adc.c:520: ERROR: [BRC_M_SMT] that open brace { should be on the previous line

**Comment by seinfra** [comment]
Target : [fcdb57c04750a0de3fa5955356e90bf9552fffa1] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fcdb57c04750a0de3fa5955356e90bf9552fffa1] - Code Rule Check OK.

File: apps/examples/testcase/ta_tc/systemio/utc/utc_adc.c (hunk 1 above)
Comment (Hyunjun85) [reply apps/examples/testcase/ta_tc/systemio/utc/utc_adc.c]: removed.

File: os/arch/arm/src/imxrt/imxrt_adc.c (hunk 2 above)
Comment (Hyunjun85) [reply os/arch/arm/src/imxrt/imxrt_adc.c]: reverted.

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 4]
```diff
@@ -83,6 +84,64 @@
...
+ *   chanlist  - The list of channels
+ *   cchannels - Number of channels
+ *
+ * Returned Value:
+ *   Valid ADC device structure reference on succcess; a NULL on failure
+ *
+ ****************************************************************************/
+void imxrt_board_adc_initialize(void)
+{
+	lldbg("%s IN\n", __FUNCTION__);
```
Comment (sunghan-chang) [inline os/board/imxrt1050-evk/src/imxrt_boot.c]: 1. I don't understand why this is necessary.
2. If this is needed indeed, use `llvdbg`. `dbg` for error case, but this is not.

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 5]
```diff
@@ -83,6 +84,64 @@
...
+#ifdef CONFIG_ANALOG
+	static bool adc_initialized = false;
```
Comment (sunghan-chang) [inline os/board/imxrt1050-evk/src/imxrt_boot.c]: Let's include <stdbool.h> to use `bool`.

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 6]
```diff
@@ -83,6 +84,64 @@
...
+	struct adc_dev_s *adc;
+	const uint8_t chanlist[] = {
```
Comment (sunghan-chang) [inline os/board/imxrt1050-evk/src/imxrt_boot.c]: <stdint.h> for uint8_t

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 7]
```diff
@@ -83,6 +84,64 @@
...
+		GPIO_PIN17,   //A4
+		GPIO_PIN20,   //A3
+		GPIO_PIN11,   //A2
+		GPIO_PIN27,   //A1
+		GPIO_PIN26    //A0
+	};
+
+	if (!adc_initialized) {
+		adc = imxrt_adc_initialize(chanlist, sizeof(chanlist) / sizeof(uint8_t));
+		if (NULL != adc) {
```
Comment (sunghan-chang) [inline os/board/imxrt1050-evk/src/imxrt_boot.c]: sys/types.h for null

Comment (Hyunjun85) [reply os/board/imxrt1050-evk/src/imxrt_boot.c]: Added.

File: os/board/imxrt1050-evk/src/imxrt_boot.c (hunk 6 above)
Comment (Hyunjun85) [reply os/board/imxrt1050-evk/src/imxrt_boot.c]: Added.

File: os/board/imxrt1050-evk/src/imxrt_boot.c (hunk 5 above)
Comment (Hyunjun85) [reply os/board/imxrt1050-evk/src/imxrt_boot.c]: Added.

File: os/board/imxrt1050-evk/src/imxrt_boot.c (hunk 4 above)
Comment (Hyunjun85) [reply os/board/imxrt1050-evk/src/imxrt_boot.c]: Removed.

**Comment by seinfra** [comment]
Target : [1b4f489a0468b076e52edeb2b4705f6cf59e194b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1b4f489a0468b076e52edeb2b4705f6cf59e194b] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [de05b3b3259098970ba487e8f073ed3b4ba32965] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [de05b3b3259098970ba487e8f073ed3b4ba32965] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [5c70df2fadfb589136d192d3d67f9c80a0f12235] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [5c70df2fadfb589136d192d3d67f9c80a0f12235] - Code Rule Check (C++) OK.

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 8]
```diff
@@ -63,18 +63,80 @@
...
+#include <sys/types.h>
+
 #include <arch/board/board.h>
 
 #include "imxrt1050-evk.h"
 #include "imxrt_start.h"
 #include "imxrt_pwm.h"
 #include "imxrt_flash.h"
 #include "imxrt_gpio.h"
+#include <tinyara/gpio.h>
```
Comment (sunghan-chang) [inline os/board/imxrt1050-evk/src/imxrt_boot.c]: Let's remove. (duplicated)

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 9]
```diff
@@ -63,18 +63,80 @@
 #include <tinyara/pwm.h>
 #include <tinyara/gpio.h>
 #include <tinyara/spi/spi.h>
+#include <tinyara/analog/adc.h>
+#include <stdbool.h>
+#include <stdint.h>
```
Comment (sunghan-chang) [inline os/board/imxrt1050-evk/src/imxrt_boot.c]: Let's remove. (duplicated)

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 10]
```diff
@@ -63,18 +63,80 @@
 #include <tinyara/pwm.h>
 #include <tinyara/gpio.h>
 #include <tinyara/spi/spi.h>
+#include <tinyara/analog/adc.h>
+#include <stdbool.h>
```
Comment (sunghan-chang) [inline os/board/imxrt1050-evk/src/imxrt_boot.c]: Please locate <stdbool.h> and <sys/types.h> as shown below:
```
#include <tinyara/config.>

#include <sys/types.h>    <--- here
#include <stdbool.h>      <--- here
#include <stdint.h>
#include <stdio.h>
```

File: os/board/imxrt1050-evk/src/imxrt_boot.c (hunk 8 above)
Comment (Hyunjun85) [reply os/board/imxrt1050-evk/src/imxrt_boot.c]: removed.

File: os/board/imxrt1050-evk/src/imxrt_boot.c (hunk 10 above)
Comment (Hyunjun85) [reply os/board/imxrt1050-evk/src/imxrt_boot.c]: relocated.

**Comment by seinfra** [comment]
Target : [3c1686846cfd2372e7bcab29c494fd34bd2d0f61] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3c1686846cfd2372e7bcab29c494fd34bd2d0f61] - Code Rule Check OK.

**Pull request merged**